File: 3297-count-substrings-that-can-be-rearranged-to-contain-a-string-i/3297-count-substrings-that-can-be-rearranged-to-contain-a-string-i.py
class Solution:
    def validSubstringCount(self, word1: str, word2: str) -> int:
        from collections import defaultdict

        counts_word2 = defaultdict(int)
        required_letters = set(word2)
        for c in word2:
            counts_word2[c] += 1

        N = len(word1)
        counts = defaultdict(int)
        total_substrings = 0
        left = 0
        right = 0

        while left < N:
            while right < N and any(counts[char] < counts_word2[char] for char in counts_word2):
                counts[word1[right]] +=1
                right +=1
            if any(counts[char] < counts_word2[char] for char in counts_word2):
                break
            total_substrings += N - right +1
            counts[word1[left]] -=1
            left +=1

        return total_substrings        

File: 3297-count-substrings-that-can-be-rearranged-to-contain-a-string-i/3297-count-substrings-that-can-be-rearranged-to-contain-a-string-i.cpp
class Solution {
public:
    long long validSubstringCount(string word1, string word2) {
        vector<int> required(26, 0);
        for(char c : word2) required[c - 'a']++;
        int requiredUnique = 0;
        for(int cnt : required) if(cnt > 0) requiredUnique++;
        vector<int> window(26);
        int formed = 0;
        long long res = 0;
        int left = 0;
        for(int right = 0; right < word1.size(); right++){
            char c = word1[right];
            window[c - 'a']++;
            if(required[c - 'a'] > 0 && window[c - 'a'] == required[c - 'a']) formed++;
            while(left <= right && formed == requiredUnique){
                res += word1.size() - right;
                char lc = word1[left];
                window[lc - 'a']--;
                if(required[lc - 'a'] > 0 && window[lc - 'a'] < required[lc - 'a']) formed--;
                left++;
            }
        }
        return res;
    }
};